### ISSUE TITLE
GPU Support

### ISSUE THREAD

**Issue by Richienb** [pavlin-policar/openTSNE#266]
openTSNE could be made GPU/CPU-agnostic for potential speed boosts by [using a cupy feature](https://stackoverflow.com/a/76514768/8384910).

**Comment by Richienb** [comment]
Switching between extending `np.ndarray` and `cupy.ndarray` may be difficult:

https://github.com/pavlin-policar/openTSNE/blob/52ae1d67cbe2b99995e6c8dc0fcc3992344998bc/openTSNE/tsne.py#L404

**Comment by Richienb** [comment]
Could use https://github.com/facebookresearch/faiss, which has gpu support.

**Comment by pavlin-policar** [comment]
Yes, FAISS could be used for NN-search and we could depend on that. However, that's just one of the three steps in t-SNE and I wouldn't want to incorporate GPU support piecemeal.

**Comment by Richienb** [comment]
> From my understanding, iterative algorithms like binary search aren't all that well suited to gpu computation

Despite this, successful GPU implementations of T-SNE still use binary search:

- https://github.com/tensorflow/tfjs-tsne/blob/bee7b5b342e4b86ec8bbb5032e606505ed0c3bce/src/tsne_optimizer_util.ts#L722 https://doi.org/10.1109/TVCG.2019.2934307
- https://github.com/CannyLab/tsne-cuda/blob/44249b6895a2eb389b8a13390ed6fb125d2040c8/src/kernels/perplexity_search.cu#L123 https://doi.org/10.1109/CAHPC.2018.8645912 https://doi.org/10.1016/j.jpdc.2019.04.008
- https://github.com/palle-k/tsne-pytorch/blob/0bf7ac9e0daee02c5cfdf4c245f6448093fd8d27/tsne_torch/tsne_torch.py#L54

Since they still achieve orders-of-magnitude performance improvement, it seems that it is only a challenge in theory.

Although it might be more theoretically sound to use some "better" algorithm, we can definitely start with just using binary search.

In my opinion, binary search is fine because GPUs can be used to train neural networks, whereby instead of using binary search to descend on weights, an even more complex (but maybe more efficient) algorithm is used for gradient descent.

**Comment by Richienb** [comment]
> I wouldn't want to incorporate GPU support piecemeal.

👍🏻

**Comment by dkobak** [comment]
I want to mention that there is also https://github.com/berenslab/contrastive-ne that implements various sampling-based t-SNE approximations (like e.g. InfoNC-t-SNE) in PyTorch. These sampling-based losses are much more amenable for GPU implementations. Personally I don't think porting Barnes-Hut or Fourier Approx to GPU is worth the effort.

**Comment by pavlin-policar** [comment]
Sure, I can reopen. While integration with pytorch would probably be on the easier side, this is not a route I'd want to go down. torch is a huge dependency and comes with all sorts of autograd and neural network stuff that would just be additional bloat here. If we add gpu support, it would be an optional dependency, but even in that case, I'd be hesitant to depend on torch. jax is much smaller, so perhaps that would be a better option. However, I've never worked with jax before, so I don't know know the limitations. From what I can tell, jax is sort of like numpy that can run on different hardware, but with an autodiff engine. Even if we don't need the autodiff engine, it's way more barebones than pytorch.

**Comment by dkobak** [comment]
To me this sounds out of scope, to be honest... @Richienb There already exist GPU versions of t-SNE. Why don't you just use one of them? There is https://github.com/berenslab/contrastive-ne from our lab based on contrastive learning in PyTorch, and also https://github.com/CannyLab/tsne-cuda that implemented FIt-SNE in CUDA.

**Comment by dkobak** [comment]
Hmm. If you have any issues running CNE with the latest PyTorch version, please do raise an issue there. Also if there is some specific functionality that you are missing in CNE, feel free to raise an issue as well.
